flang/include/flang/Lower/Support/BoxValue.h

```diff
@@ -382,6 +399,24 @@ class ExtendedValue : public details::matcher<ExtendedValue> {
     return getBoxOf<UnboxedValue>();
   }
 
+  /// If the entity in the box does not have a contiguous memory layout,
+  /// returns the fir.box describing the memory layout, otherwise, returns
+  /// the base.
+  mlir::Value getMemRef() const {
+    return match(
+        [](const ArrayBoxValue &box) {
+          return box.isContiguous() ? box.getAddr() : box.getSourceBox();
+        },
+        [](const CharArrayBoxValue &box) {
+          return box.isContiguous() ? box.getAddr() : box.getSourceBox();
+        },
+        [](const BoxValue &box) {
+          return box.isContiguous() ? box.getAddr() : box.getSourceBox();
+        },
+        [](const UnboxedValue &unboxed) { return unboxed; },
+        [](const auto &box) { return box.getAddr(); });
+  }
+
```

Comment:
I'm a little confused.

An extended array value can come from any expression. The CONTIGUOUS Fortran attribute is really more about function arguments, no? That is, a local array is necessarily contiguous by allocation. But a boxed array can be a projection, need not be contiguous, but might be flagged/asserted as contiguous at the syntactic layer.

Reply:
The `box.isContiguous()` is not intended to tell "Does this entity has the CONITGUOUS" keyword, it is intended to tell: "Does this ExtendedValue describes an entity that is known to be contiguous at compile time."
An ExtendedValue know to be contiguous at compile time is:
- a temp
- a simply contiguous designator (defined in section 9.5.4)

So a Designator expression of a dummy argument with the CONTIGUOUS attribute will lower to an ExtendedValue for which `isContiguous` returns true, but the reciprocal is not true (i.e., `isContiguous`  dos not imply the extended value describes a dummy with the CONTIGUOUS attribute).

Right now `isContiguous` may not implement the whole "is simply contiguous" logic, but that is really the intension.
I am assuming expressions that are not simply contiguous and need to be represented as Extended Values (e.g. entities associated to an array section) will have a fir.box generated by an embox with a slice (and that would be the soureBox for the ExtendedValue).